File: Template/Naked Functions Server/Template.Server/Startup.cs
// Copyright Naked Objects Group Ltd, 45 Station Road, Henley on Thames, UK, RG9 1AT
// Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License.
// You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
// Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and limitations under the License.

using Microsoft.AspNetCore.Builder;
using Microsoft.AspNetCore.Hosting;
using Microsoft.Extensions.Configuration;
using Microsoft.Extensions.DependencyInjection;
using Microsoft.Extensions.Hosting;
using Microsoft.Extensions.Logging;
using NakedFramework;
using NakedFramework.Architecture.Component;
using NakedFramework.DependencyInjection.Extensions;
using NakedFramework.Persistor.EFCore.Extensions;
using NakedFramework.Rest.Extensions;
using NakedFunctions.Reflector.Extensions;
using Newtonsoft.Json;
using Template.Model;

namespace NakedFunctions.Rest.App.Demo
{
    public class Startup
    {

        readonly string MyAllowSpecificOrigins = "_myAllowSpecificOrigins";

        public Startup(IConfiguration configuration)
        {
            Configuration = configuration;
        }

        public IConfiguration Configuration { get; }


        // This method gets called by the runtime. Use this method to add services to the container.
        public void ConfigureServices(IServiceCollection services)
        {
            services.AddControllers()
                .AddNewtonsoftJson(options => options.SerializerSettings.DateTimeZoneHandling = DateTimeZoneHandling.Utc);
            services.AddMvc(options => options.EnableEndpointRouting = false);
            services.AddHttpContextAccessor();
            services.AddNakedFramework(builder =>
            {
                builder.MainMenus = MenuHelper.GenerateMenus(ModelConfig.MainMenus());
                builder.AddEFCorePersistor(options => { options.ContextInstallers = new[] { ModelConfig.EFCoreDbContextInstaller }; });
                builder.AddNakedFunctions(options =>
                {
                    options.FunctionalTypes = ModelConfig.Types();
                    options.Functions = ModelConfig.Functions();
                });
                builder.AddRestfulObjects(_ => { });
            });
            services.AddCors(options =>
            {
                options.AddPolicy(MyAllowSpecificOrigins, builder =>
                {
                    builder
                        .WithOrigins("http://localhost:5001")
                        .AllowAnyHeader()
                        .WithExposedHeaders("Warning", "ETag", "Set-Cookie")
                        .AllowAnyMethod()
                        .AllowCredentials();
                });
            });
        }

        // This method gets called by the runtime. Use this method to configure the HTTP request pipeline.
        public void Configure(IApplicationBuilder app, IWebHostEnvironment env, IModelBuilder builder, ILoggerFactory loggerFactory)
        {

            // for Demo use Log4Net. Configured in log4net.config  
            loggerFactory.AddLog4Net();

            builder.Build();

            if (env.IsDevelopment())
            {
                app.UseDeveloperExceptionPage();
            }

            app.UseCors(MyAllowSpecificOrigins);
            app.UseRouting();
            app.UseRestfulObjects();
        }
    }
}

File: Template/Naked Functions Server/Template.Server.Fsharp/Startup.fs
// Copyright Naked Objects Group Ltd, 45 Station Road, Henley on Thames, UK, RG9 1AT
// Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License.
// You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
// Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and limitations under the License.

namespace NakedFunctions.Rest.App.Demo

open System
open Microsoft.AspNetCore.Authentication.JwtBearer
open Microsoft.AspNetCore.Builder
open Microsoft.AspNetCore.Hosting
open Microsoft.EntityFrameworkCore
open Microsoft.Extensions.Configuration
open Microsoft.Extensions.DependencyInjection
open Microsoft.Extensions.Hosting
open Microsoft.Extensions.Logging
open NakedFramework.Architecture.Component
open NakedFramework.DependencyInjection.Extensions
open NakedFramework.Menu
open NakedFramework.Persistor.EFCore.Extensions
open NakedFramework.Rest.Extensions
open NakedFunctions.Reflector.Extensions
open Newtonsoft.Json
open Template.Model.Fsharp

type Startup(configuration: IConfiguration) =
    let Configuration = configuration

    member this.MyAllowSpecificOrigins = "_myAllowSpecificOrigins"

    // This method gets called by the runtime. Use this method to add services to the container.
    member this.ConfigureServices(services: IServiceCollection) =

        services
            .AddAuthentication(fun options ->
                options.DefaultAuthenticateScheme <- JwtBearerDefaults.AuthenticationScheme
                options.DefaultChallengeScheme <- JwtBearerDefaults.AuthenticationScheme)
            .AddJwtBearer(fun options ->
                let domain = "Auth0:Domain"
                options.Authority <- $"https://{Configuration.[domain]}/"
                options.Audience <- Configuration.["Auth0:Audience"]

                options.TokenValidationParameters.NameClaimType <-
                    "http://schemas.xmlsoap.org/ws/2005/05/identity/claims/emailaddress")
        |> ignore

        services
            .AddControllers()
            .AddNewtonsoftJson(fun options ->
                options.SerializerSettings.DateTimeZoneHandling <- DateTimeZoneHandling.Utc)
        |> ignore

        services.AddMvc(fun options -> options.EnableEndpointRouting <- false)
        |> ignore

        services.AddHttpContextAccessor() |> ignore

        services.AddNakedFramework
            (fun frameworkOptions ->
                frameworkOptions.MainMenus <- MenuHelper.GenerateMenus(ModelConfig.MainMenus())

                frameworkOptions.AddEFCorePersistor
                    (fun peristorOptions ->
                        peristorOptions.ContextCreators <-
                            [| Func<IConfiguration, DbContext> ModelConfig.EFCoreDbContextCreator |])

                frameworkOptions.AddNakedFunctions
                    (fun appOptions ->
                        appOptions.DomainTypes <- ModelConfig.DomainTypes()
                        appOptions.DomainFunctions <- ModelConfig.TypesDefiningDomainFunctions())

                frameworkOptions.AddRestfulObjects(fun _ -> ()))

        services.AddCors
            (fun corsOptions ->
                corsOptions.AddPolicy(
                    this.MyAllowSpecificOrigins,
                    fun policyBuilder ->
                        policyBuilder
                            .WithOrigins("http://localhost:5001")
                            .AllowAnyHeader()
                            .WithExposedHeaders("Warning", "ETag", "Set-Cookie")
                            .AllowAnyMethod()
                            .AllowCredentials()
                        |> ignore
                ))
        |> ignore



    // This method gets called by the runtime. Use this method to configure the HTTP request pipeline.
    member this.Configure
        (app: IApplicationBuilder)
        (env: IWebHostEnvironment)
        (builder: IModelBuilder)
        (loggerFactory: ILoggerFactory)
        =
        // for Demo use Log4Net. Configured in log4net.config
        loggerFactory.AddLog4Net() |> ignore

        builder.Build()

        if (env.IsDevelopment()) then
            app.UseDeveloperExceptionPage() |> ignore

        app
            .UseAuthentication()
            .UseCors(this.MyAllowSpecificOrigins)
            .UseRouting()
            .UseRestfulObjects()
